cli/Sources/TuistServer/Client/ServerAuthenticationController.swift

```diff
@@ -64,129 +213,284 @@ public struct ServerAuthenticationController: ServerAuthenticationControlling {
+            return .valid(.user(
+                accessToken: accessToken,
+                refreshToken: refreshToken
+            ))
+        }
+    }
+
+    func executeRefresh(serverURL: URL, forceRefresh: Bool) async throws -> (value: AuthenticationToken, expiresAt: Date?)? {
+        guard let token = try await fetchTokenFromStore(serverURL: serverURL) else {
+            return nil
+        }
+
+        let upToDateToken: AuthenticationToken
+        var expiresAt: Date?
+
+        switch token {
+        case .project:
+            upToDateToken = token
+        case let .user(
+            accessToken: accessToken, refreshToken: refreshToken
+        ):
+            // We consider a token to be expired if the expiration date is in the past or 30 seconds from now
+            let now = Date.now()
+            let expiresIn = accessToken.expiryDate
+                .timeIntervalSince(now)
+            let refresh = expiresIn < 30 || forceRefresh
+
+            #if canImport(TuistSupport)
+                if refresh {
+                    Logger.current.debug(
+                        "Access token expires in less than \(expiresIn) seconds. Renewing..."
+                    )
+                } else {
+                    Logger.current.debug(
+                        "Access token expires in \(expiresIn) seconds and it is still valid"
+                    )
+                }
+            #endif
+            if refresh {
                 #if canImport(TuistSupport)
-                    Logger.current
-                        .warning(
-                            "You are using a deprecated user token. Please, reauthenticate by running 'tuist auth login'."
-                        )
+                    Logger.current.debug("Refreshing access token for \(serverURL)")
                 #endif
-                return .user(
-                    legacyToken: $0.token,
-                    accessToken: nil,
-                    refreshToken: nil
+                let tokens = try await refreshTokens(
+                    serverURL: serverURL, refreshToken: refreshToken
                 )
+                #if canImport(TuistSupport)
+                    Logger.current.debug("Access token refreshed for \(serverURL)")
+                #endif
+                upToDateToken = .user(
+                    accessToken: try JWT.parse(tokens.accessToken),
+                    refreshToken: try JWT.parse(tokens.refreshToken)
+                )
+                expiresAt = try JWT.parse(tokens.accessToken)
+                    .expiryDate
+            } else {
+                upToDateToken = .user(
+                    accessToken: accessToken,
+                    refreshToken: refreshToken
+                )
+                expiresAt = accessToken.expiryDate
             }
         }
+        return (value: upToDateToken, expiresAt: expiresAt)
+    }
+
+    #if canImport(TuistSupport)
+        public func lockFilePath(serverURL: URL) -> AbsolutePath {
```

Comment:
```suggestion
        private func lockFilePath(serverURL: URL) -> AbsolutePath {
```

I don't think it needs to be public, does it ... ? I see it used only in this file